src/natcap/invest/coastal_blue_carbon/coastal_blue_carbon.py

```diff
@@ -188,941 +187,1485 @@
+        (min(transition_years) - 1): {
+            POOL_SOIL: args['annual_rate_of_accumulation_rasters'][POOL_SOIL],
+            POOL_BIOMASS: (
+                args['annual_rate_of_accumulation_rasters'][POOL_BIOMASS]),
+            POOL_LITTER: (
+                args['annual_rate_of_accumulation_rasters'][POOL_LITTER]),
+        }
+    }
+    disturbance_vol_rasters = {}
+    emissions_rasters = {}
+    year_of_disturbance_rasters = {}
+    total_carbon_rasters = {}
+    prior_transition_year = None
+    current_transition_year = None
+
+    current_disturbance_vol_tasks = {}
+    prior_stock_tasks = {}
+    current_year_of_disturbance_tasks = {}
+    current_emissions_tasks = {}
+    prior_net_sequestration_tasks = {}
+    current_net_sequestration_tasks = {}
+    valuation_tasks = {}
+
+    first_transition_year = min(transition_years)
+    final_year = int(args['analysis_year'])
+
+    summary_net_sequestration_tasks = []
+    summary_net_sequestration_raster_paths = {
+        first_transition_year: args['sequestration_since_baseline_raster']}
+
+    for year in range(first_transition_year, final_year+1):
+        current_stock_tasks = {}
+        net_sequestration_rasters[year] = {}
+        stock_rasters[year] = {}
+        disturbance_vol_rasters[year] = {}
+        emissions_rasters[year] = {}
+        year_of_disturbance_rasters[year] = {}
+        valuation_tasks[year] = {}
+
+        for pool in (POOL_SOIL, POOL_BIOMASS):
+            # Calculate stocks from last year's stock plus last year's net
+            # sequestration.
+            stock_rasters[year][pool] = os.path.join(
+                intermediate_dir,
+                STOCKS_RASTER_PATTERN.format(
+                    year=year, pool=pool, suffix=suffix))
+            if year == first_transition_year:
+                current_stock_dependent_tasks = []
+                current_disturbance_vol_dependent_tasks = []
+            else:
+                current_stock_dependent_tasks = [
+                    prior_stock_tasks[pool],
+                    prior_net_sequestration_tasks[pool]]
+                current_disturbance_vol_dependent_tasks = [
+                    prior_stock_tasks[pool]]
+
+            current_stock_tasks[pool] = task_graph.add_task(
+                func=_sum_n_rasters,
+                args=([stock_rasters[year-1][pool],
+                       net_sequestration_rasters[year-1][pool]],
+                      stock_rasters[year][pool]),
+                dependent_task_list=current_stock_dependent_tasks,
+                target_path_list=[stock_rasters[year][pool]],
+                task_name=f'Calculating {pool} carbon stock for {year}')
+
+            # Calculate disturbance volume if we're at a transition year.
+            if year in transition_years:
+                # We should only switch around the transition years the first
+                # time we encounter this, not for each pool.
+                if current_transition_year != year:
```

Comment:
I'm a little confused about what the comment is saying here vs. what it's doing. It looks like `current_transition_year` will get updated on each iteration, because `year` will increment each time. And it makes sense that you're keeping track of the prior and current transition years on lines 775 and 776. But I'm not sure why line 774 is necessary (shouldn't it always be true?) and the comment seems to contradict that. Please correct me if I'm missing something :)

Reply:
Ha, this check was the result of a neat bug in bookkeeping! 

An abbreviated form of the timeseries loop up to this conditional would be:
```python
for year in range(first_transition_year, final_year+1):
    # ...
    for pool in (POOL_SOIL, POOL_BIOMASS):
        # ...
        if year in transition_years:
            # ...

            # We should only switch around the transition years the first
            # time we encounter this, not for each pool.
            if current_transition_year != year:
                prior_transition_year = current_transition_year
                current_transition_year = year
```

If we don't have the `if current_transition_year != year`, the lines
```python
prior_transition_year = current_transition_year
current_transition_year = year
```
would be executed twice in the matching year, once each for `POOL_SOIL` and `POOL_BIOMASS`, resulting in both `prior_transition_year` and `current_transition_year` pointing to the same year, which we don't want.  The condition here ensures that we only update the prior/current transition years the first time we encounter it.

So if we're visiting `if current_transition_year != year` the first time, you're right that the two variables, `current_transition_year` and `year` will not have the same value.  So then we execute the next lines, setting `prior_transition_year`, and then `current_transition_year` is now set to `year.  So then for the next carbon pool, the next time we reach line 774, ` current_transition_year` will have already been set to `year`, and we don't change our understanding of the prior/current year.

I do think it would be clearer to have this logic elsewhere in the timeseries loop so that we could avoid this possible duplication, but I'm not convinced that this could be done without duplicating conditionals elsewhere as well.  I'm happy to revisit this if you like ... let me know what you think!

Reply:
Ohh, that makes sense! I had missed the `for pool in (POOL_SOIL, POOL_BIOMASS)` loop. Thanks for explaining!